### ISSUE TITLE
Z-ordering issue in 3D and 2D maps

### ISSUE THREAD

**Issue by Metibor** [Pyrdacor/Ambermoon.net#3]
Apart from the palette problem I described in the other issue, I am having Z-ordering problems in both map types.
Your screenshots suggest, that you don't have the problem, so I will post it as an issue.

This is the room with the well in the first 3D map which is mostly looking okay:
![grafik](https://user-images.githubusercontent.com/12674912/90164890-3c9e5b80-dd98-11ea-8d82-07436d9a0541.png)

But if I turn around all walls and objects are drawn in the wrong order:
![grafik](https://user-images.githubusercontent.com/12674912/90164799-20022380-dd98-11ea-9ba8-bc6c6a6ec10d.png)

In 2D maps I have the issue, that the main character is drawn behind the decoration tiles
![grafik](https://user-images.githubusercontent.com/12674912/90165518-11683c00-dd99-11ea-9b99-6452fbe6a6db.png)

![grafik](https://user-images.githubusercontent.com/12674912/90165647-42e10780-dd99-11ea-9baa-6af1d2768710.png)

I am using a Radeon HD 5670 on one computer and a HD6850 on my other. Both have the same problem.

Btw, I am closely following the project since day 2 or 3.
I will try to help where I can, but I don't have much experience in reverse engineering!

**Comment by Pyrdacor** [comment]
Thanks for the report. I don't have that issues. I will have a look into it soon.

**Comment by Metibor** [comment]
I have an update on this issue with a partial solution:

If I set the preferredDepthBufferBits to an explicit value like this:
```
            var options = new WindowOptions(true, true, new System.Drawing.Point(100, 100),
                new System.Drawing.Size(Width, Height), 60.0, 60.0, GraphicsAPI.Default,
                $"Ambermoon.net v{version.Major}.{version.Minor}.{version.Build}",
                WindowState.Normal, WindowBorder.Fixed, VSyncMode.Off,
                10, false, videoMode, 24);
```
the 2D and 3D maps render as expected.

![grafik](https://user-images.githubusercontent.com/12674912/90403715-a3c25580-e0a1-11ea-9604-738137fafa1c.png)

![grafik](https://user-images.githubusercontent.com/12674912/90403559-74abe400-e0a1-11ea-92f2-23a8b3ff2be4.png)

The problem is now, that it only works for the first 3D map. 
On visiting another map, i.e. Spannenberg after Großvaters Keller or vice versa I get horrible flickering and wrong graphic elements.
![grafik](https://user-images.githubusercontent.com/12674912/90403953-fc91ee00-e0a1-11ea-81ba-f8dab2a25ef3.png)

![grafik](https://user-images.githubusercontent.com/12674912/90404133-3bc03f00-e0a2-11ea-99c8-14ff3c65e493.png)

I'm pretty sure, that the depth buffer (or something else) needs to be reinitialized somewhere (but I don't know how yet).

With the depth buffer bits set I also get some glitches in the world map sometimes:
![grafik](https://user-images.githubusercontent.com/12674912/90405202-c190ba00-e0a3-11ea-91dc-47b47703902e.png)

They even change position while wandering around:
![grafik](https://user-images.githubusercontent.com/12674912/90405095-9c9c4700-e0a3-11ea-9d19-d1059377a917.png)

On indoor 2D maps, i.e. Großvaters Haus, the glitch does not show up or it goes away when entering an indoor map from the world map, where the glitch is shown.

Maybe you have an idea, what is happening now?

**Comment by Pyrdacor** [comment]
Hm that is strange. Maybe some settings of the display adapter that interferes with something I used? Normally AMD cards are closer to the Opengl specs than nvidia ones so maybe I did something which is not intended and nvidia just manages it to work.

I use Silk.net as an opengl backend which uses glfw for window creation. There the depth bits come into play. I myself added that option to silk.net for the freeserf project. But it the end it wasn't needed there. But the Ambermoon engine is based on that of freeserf.

It could be some issues with glfw or Silk.net. In general an appropriate depth buffer bit size should be chosen by default if available (mostly 24 bits). Strange that it does not on your end.

I would check your adapter settings or check if glfw/silk has some issues with depth/and. I have no other idea at the moment.

**Comment by Pyrdacor** [comment]
I checked silk and glfw code. The glfw 3 specs state that the default value for the depth buffer bits is 24. But the implementation is different. At first all possible modes are queried. Then the best match is searched. But if no preference for the depth buffer bits is given, any mode could be chosen. So it is correct and important to set the preferred depth bits to 24. I will do so soon.

But this won't fix your glitches. I am not sure what causes them. I know that the 3D map cleanup is not implemented yet so it could be that there are relicts from the last map. This could also affect 2D maps. But 2D maps should clean up their stuff. But maybe in some circumstance they won't. Do the 2D glitches also occur without entering any 3D map before?

**Comment by Pyrdacor** [comment]
I stand corrected. The 3D cleanup code was in there but when entering a new 3D map, a totally new instance was created and the old one wasn't touched. I fixed it and the glitches should be gone now.

**Comment by Metibor** [comment]
3D maps are working perfect now... as perfect as they are for now. 😉

I just had to add this code to `RenderBuffer.FreeDrawIndex(int index)`
```
            if (billboardOrientationBuffer != null)
            {
                billboardOrientationBuffer.Remove(index);
                billboardOrientationBuffer.Remove(index + 1);
                billboardOrientationBuffer.Remove(index + 2);
                billboardOrientationBuffer.Remove(index + 3);
            }
 ```
or else it crashes on the second 3D map.
Made a PR for it.

The sad news is, that the 2D glitch is still present.

**Comment by Pyrdacor** [comment]
I can't test this as it doesn't occur for me but a shot in the blue would be that the current frame of a tile is greater than 0 cause of an animation. When you then switch the map, the tiles will be reused in 2d but with new images. The current frame might stay at its value and therefore displays some strange image. From a fast look into the code from my mobile phone it might help to set the CurrentFrame to 0 in UpdateTiles. But it's only a guess.

**Comment by Pyrdacor** [comment]
Do the 2d glitches stay at the same position relative to the screen when you move?

**Comment by Pyrdacor** [comment]
From your screenshots it looks like the artefacts have the correct shape so my first guess would be that either the palette index or the color index is just wrong. Maybe the precision issue inside the shader again? But the fix with "+ 0.5f" should be save. I tested the resulting values and they should never overlap in any way. Do you use a modified texture shader?

I need more info when and where this happens and if the affected tiles are always the same (either in relation to the map or in relation to the screen/window).

**Comment by Pyrdacor** [comment]
@Metibor FYI I re-added the libraries to this repo. It was a pain in the *** to debug changes. I will remove them in the other repo as they will be available as nuget packages anyway and there will only be 1 or 2 small tools which use them. The main focus is now on the game which will also push the development of the libs so I think this is the better approach.

**Comment by Metibor** [comment]
Current facts:
- I cannot reproduce it reliable for now
- If it happens, I get mostly two glitched tiles on random positions (but have seen more than two) on Lyramion
- The glitched tiles are relative to the player position (if the player moves, they follow)
- They show up only on tiles with more than one layer
- If the player moves and a glitched tile would end up on a tile with only one layer (i.e. grass) it vanishes and after the next move it shows up on another multi layered tile
- I use the current shader 
- It stops if I enter any other map (2D or 3D)

Here is a series of screenshots:
The two tiles are connected, which is random.
I walk to the right until they move on (or under?) the grass
After that I walk to the left and they show up at another position (still connected, but this is random again)
![grafik](https://user-images.githubusercontent.com/12674912/90555633-f7af6600-e197-11ea-87fe-37372597c1d8.png)
![grafik](https://user-images.githubusercontent.com/12674912/90555671-0433be80-e198-11ea-95c6-cd4b2aae9a56.png)
![grafik](https://user-images.githubusercontent.com/12674912/90555679-08f87280-e198-11ea-9a6b-9badf82baad8.png)
![grafik](https://user-images.githubusercontent.com/12674912/90555697-10b81700-e198-11ea-8994-cf7144213ebf.png)
![grafik](https://user-images.githubusercontent.com/12674912/90555711-157ccb00-e198-11ea-9b3b-ab56620f1d47.png)
![grafik](https://user-images.githubusercontent.com/12674912/90555725-1c0b4280-e198-11ea-88de-bfe72fd30edb.png)
![grafik](https://user-images.githubusercontent.com/12674912/90555746-23cae700-e198-11ea-8861-f229bf9d3da9.png)
![grafik](https://user-images.githubusercontent.com/12674912/90555840-4230e280-e198-11ea-8cf2-97a9cce4d1e9.png)

So it seems to be (another?) z-buffer problem, but I need to investigate further and see how it can be reproduced.

Regarding the libs: 😎

**Comment by Thallyrion** [comment]
If I may add, I never noticed such a glitch.

Am Di., 18. Aug. 2020 um 21:49 Uhr schrieb Jens <notifications@github.com>:

**Comment by Pyrdacor** [comment]
The black fragments are definitely the front tile graphics as back tiles do not have fully transparent parts and therefore would be perfect squares. It looks more like a blending issue.

**Comment by Metibor** [comment]
Nothing much to add here for now, other than that it seems to be an AMD thing.
Could not reproduce it on other computers, but they all have Nvidia GPUs.

I still don't know, when it happens.
But if it happens, then it shows up on the upper right corner of the world map

Steps are:
* Start game (initial save game)
* Exit Großvaters Haus
![grafik](https://user-images.githubusercontent.com/12674912/91641770-0e22b080-ea27-11ea-8814-ac61554c7df7.png)
* Take one step south
![grafik](https://user-images.githubusercontent.com/12674912/91641784-1975dc00-ea27-11ea-942a-a29eddd76eba.png)
* Partial black square at the top right

* Reentering Großvaters Haus
* Exiting Großvaters Haus again
* No glitched tile

But it may happen again later and with more than one tile, when exiting other 2D or 3D maps.


Another thing that happens sometimes:
I get random Exceptions (mostly on map changes) in line 937 and 939 of RenderBuffer.

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at Silk.NET.OpenGL.X96E42541.GL.DrawElements(Silk.NET.OpenGL.PrimitiveType, UInt32, Silk.NET.OpenGL.DrawElementsType, Void*)
   at Ambermoon.Renderer.RenderBuffer.Render()
   at Ambermoon.Renderer.RenderLayer.Render()
   at Ambermoon.Renderer.OpenGL.RenderView.Render()
   at Ambermoon.GameWindow.Window_Render(Double)
   at Silk.NET.Windowing.Desktop.GlfwWindow.RaiseRenderFrame()
   at Silk.NET.Windowing.Desktop.GlfwWindow.Run(System.Action)
   at Silk.NET.Windowing.Common.WindowExtensions.Run(Silk.NET.Windowing.Common.IView)
   at Ambermoon.GameWindow.Run(Ambermoon.Configuration)
   at Ambermoon.Program.Main()
```

```csharp 
                try
                {
                    if (positionBuffer != null)
/*937*/                 state.Gl.DrawElements(PrimitiveType.Triangles, (uint)(positionBuffer.Size / 4) * 3, DrawElementsType.UnsignedInt, (void*)0);
                    else if (vectorBuffer != null)
/*939*/                 state.Gl.DrawElements(PrimitiveType.Triangles, (uint)vectorBuffer.Size / 2, DrawElementsType.UnsignedInt, (void*)0);
                    else
                        throw new AmbermoonException(ExceptionScope.Render, "Neither position nor vector buffer exists.");
                }
                catch
                {
                    // ignore for now
                }
                finally
                {
                    vertexArrayObject.Unlock();
                }
```

I don't know if that is in some way related to the first problem.

**Comment by Pyrdacor** [comment]
This is odd. The two problems might be related. Maybe your video adapters have some OpenGL issues? I can't reproduce this either but I only have Nvidia and Intel here for testing and both work as expected.

**Comment by Pyrdacor** [comment]
Do you know RenderDoc? It's an open source tool to debug rendering. Maybe you can try it out?

**Comment by Metibor** [comment]
Wow, this looks complicated. I will look into it, but it will take some time.

Another idea: Do you have an overview, which (gaming) projects use the OpenGL part of Silk.NET?
I could have a look at those and check if I run into similar problems.

Although my guess would still be, that some part of OpenGL is not properly initialized.
Via Google I found this [question](https://stackoverflow.com/questions/42777833/opengl-gldrawelements-gives-unhandled-exception-system-accessviolationexceptio) which goes into the same direction.

**Comment by Pyrdacor** [comment]
Sure try my Settlers clone: It has nearly the same engine: github.com/Pyrdacor/freeserf.net.

The linked question lacks an index buffer. This is not the case here. You also have good rendering most of the time so this is something different. My guess is the discard in the fragment shader. I will probably improve the alpha test anyway soon.

But RenderDoc might be useful to find the issue.

**Comment by Metibor** [comment]
I found something with RenderDoc!
When leaving Spannenberg (loaded a savegame in town):

The number of glDrawElements for the backgroundTileSprites is 594 (99 tiles * 6 vertices)
![grafik](https://user-images.githubusercontent.com/12674912/92605081-93248a00-f2b1-11ea-8e0c-b2b476406838.png)

The number of glDrawElements for the foregroundTileSprites is 408 (68 tiles * 6 vertices)
![grafik](https://user-images.githubusercontent.com/12674912/92605000-7b4d0600-f2b1-11ea-94df-5e36b6d50345.png)

All tiles have paletteIndex = 0 in VS Input and palIndex = 0.00 in VS Output (Worldmap = 0)

When moving the number of foregroundTileSprites changes. If there are less foreground tiles everything is okay, but when the number increases (up to the maximum of 594) there is undefined behaviour.
![grafik](https://user-images.githubusercontent.com/12674912/92605324-e1d22400-f2b1-11ea-9fbd-5cf8e32d3eb0.png)

Every new vertex is added with an undefined paletteIndex in VS Input, which causes arbitrary values for palIndex in VS Output, that lead to the wrong colors (at least with my gpu).
![grafik](https://user-images.githubusercontent.com/12674912/92601559-45a61e00-f2ad-11ea-8fc7-0ba5a72f85ee.png)

Btw, the behaviour is so random, that even RenderDoc renders the frames sometimes with and sometimes without the error. 🤪

I hope, that you are able to locate the bug in the code, because I could not fix it.

**Comment by Pyrdacor** [comment]
This is very strange. The code for background and foreground sprites are almost identical. And the difference is mostly the foreground sprite can be changed by map events (which is not true for the world map in front of Spannenberg). That the paletteIndex is undefined is ever more strange. I would understand if you would have issues everywhere but only for foreground map sprites? Every shader uses palette indices this way and they all work.

Is your GPU in some way limited? Otherwise I would say this might be a driver bug. I can't find anything in code which could cause undefined paletteIndex inputs. And if so what would the shader do with such input? Such a thing as "undefined input" shouldn't even exist imho so I think even more this is a driver bug. Does this happen on another GPU as well?

**Comment by Pyrdacor** [comment]
I'm not familiar with RenderDoc. Maybe you can file an issue there (https://github.com/baldurk/renderdoc/issues) and ask what this "---" means and how it could happen?

**Comment by Metibor** [comment]
**YES**, I got it!

![grafik](https://user-images.githubusercontent.com/12674912/92705782-45149280-f354-11ea-9a73-9e2c4622d397.png)

The base clase if the IndexBuffer, BufferObject is not set to changed when the buffer size changes and the default value of 0 is inserted!
That is why I never had the problem on other 2D maps, because they use another paletteIndex (i.e. index = 6) than the world map (index = 0).

I made a pull request!
